PolyEngine/Core/Src/pe/core/utils/Result.hpp

```diff
@@ -0,0 +1,117 @@
+#pragma once
+
+#include <pe/Defines.hpp>
+#include <variant>
+
+namespace pe::core::utils
+{
+	// 
+	template <typename T, typename E>
+	class Result
+	{
+	public:
+		Result() = delete;
+		Result(Result&) = default;
+		Result(Result&&) = default;
+		Result& operator=(Result&) = default;
+		Result& operator=(Result&&) = default;
+		Result(T okVal) : m_value(std::move(okVal)) {}
+		Result(E errVal) : m_value(std::move(errVal)) {}
+		
+		bool isOk() const { return m_value.index() == 0; }
+		bool isErr() const { return !isOk(); }
+		
+		T& getValue() { return std::get<T>(m_value); }
+		E& getError() { return std::get<E>(m_value); }
+		const T& getValue() const { return std::get<T>(m_value); }
+		const E& getError() const { return std::get<E>(m_value); }
+
+		template <typename U>
+		Result<U, E> and(Result<U, E> res)
+		{
+			if (isOk())
+				return res;
+			else
+				return getError();
+		}
+
+		template <typename F>
+		Result<T, F> or(Result<T, F> res)
+		{
+			if (isOk())
+				return getValue();
+			else
+				return res;
+		}
+
+	private:
+		std::variant<T, E> m_value;
+	};
+	
+	template <typename E>
+	class Result<void, E> final : public BaseObjectLiteralType<>
+	{
+	public:
+		Result() = default;
+		Result(Result&) = default;
+		Result(Result&&) = default;
+		Result& operator=(Result&) = default;
+		Result& operator=(Result&&) = default;
+		Result(E errVal) : m_value(std::move(errVal)) {}
+
+		bool isOk() const { return m_value.has_value() == false; }
+		bool isErr() const { return !isOk(); }
+
+		E& getError() { return m_value.value(); }
+		const E& getError() const { return m_value.value(); }
+
+		template <typename U>
+		Result<U, E> and(Result<U, E> res)
+		{
+			if (isOk())
+				return res;
+			else
+				return getError();
+		}
+
+		template <typename F>
+		Result<void, F> or (Result<void, F> res)
+		{
+			if (isOk())
+				return Result<void, F>();
+			else
+				return res;
+		}
+
+	private:
+		std::optional<E> m_value;
+	};
+
+	template <typename T, typename E>
+	Result<T, E> join(Result<Result<T, E>, E> res)
+	{
+		if (res.isOk())
+			return res.getValue();
+		else
+			return res.getError();
+	}
+
+	template <typename E>
+	Result<void, E> join(Result<Result<void, E>, E> res)
```

Comment:
Are `join` and `bind` widely used names or are we free to name those functions as we want? I would suggest `collapse` instead of `join`. WDYT?
Also, this as well might be a simple method.

Reply:
FYI there is a std::bind

Reply:
`join` and `bind` are widely used unless someone is trying to call monads [Warm and fuzzy things](https://www.urbandictionary.com/define.php?term=Warm%20Fuzzy%20Thing) insetead

Reply:
`std::bind` is completly different thing.
I don't think it should be a simple method. Once `c++20` is a thing, it won't be long before someone somewhere (most probably in boost) defines monad concept. We should use it then

Reply:
we can have namespace result for free functions for result

Reply:
What is the use case for this function? I don't see a situation where you can end with a Result with Result inside, as this would mean that you didn't handle the error which you should have handled.

Reply:
resolved - i've removed join

Reply:
The idea here was to have this as a part of monadic interface, but true, this is practically less usefull than bind